### PR TITLE
#6 Migrate from compilerflasher to using codebender's builder service

### PR DESCRIPTION
**Pull request by kyleoliveira**
Squashes the commits from #9 cleanly. Use this pull request instead.

### CODE REVIEW

File: app/Blocklyduino.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+     * @param array $values The parameters or objects.
+     */
+    function __construct(array $values = array()) {
+        parent::__construct();
+
+        // Set to false if you don't want debug messages
+        $this['debug'] = true;
+
+        // The URL to use to connect to the Codebender Builder service
+        $this['builder_url'] = "http://builder.codebender.cc:8080/compile?api_key=blocklyduino";
```
Comment (fpapadopou) [inline app/Blocklyduino.php]: This should definitely be moved out of here. We need to use some parameters file.
In order to do that, use [this Silex component](https://github.com/deralex/YamlConfigServiceProvider). You will find everything you need in the YamlConfigServiceProvider documentation.
In fact, you should create and commit a dummy config file, and all your real parameters will be manually changed in real installations.
For example, you can create a config.yml file containing this

``` yaml
application:
    ...
    builder: http://url/to/builder.com
    ...
```

and in any real installation, you'll have to manually edit the builder parameter in the file.

File: app/Blocklyduino.php [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        // An example JSON request for testing the service connection
+        $this['debug_code_request'] = <<<HDOC
+{"files":[{"filename":"Blink Example.ino","content":"\/*\\n\\tBlink\\n\\tTurns on an LED on for one second, then off for one second, repeatedly.\\n\\n\\tThis example code is in the public domain.\\n*\/\\n\\nvoid setup()\\n{\\n\\t\/\/ initialize the digital pin as an output.\\n\\t\/\/ Pin 13 has an LED connected on most Arduino boards:\\n\\tpinMode(13, OUTPUT);\\n}\\n\\nvoid loop()\\n{\\n\\tdigitalWrite(13, HIGH); \/\/ set the LED on\\n\\tdelay(1000); \/\/ wait for a second\\n\\tdigitalWrite(13, LOW); \/\/ set the LED off\\n\\tdelay(1000); \/\/ wait for a second\\n}\\n"}],"libraries":[],"logging":true,"format":"binary","version":"105","build":{"mcu":"atmega328p","f_cpu":"16000000L","core":"arduino","variant":"standard"}}
```
Comment (fpapadopou) [inline app/Blocklyduino.php]: The default Arduino code could also be moved to a file out of your php code. You could have the filename in your config file and read it's contents at this point.

File: src/services/CodebenderServiceProvider.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+use GuzzleHttp\Client;
+use Silex\Application;
+use Silex\ServiceProviderInterface;
+
+class CodebenderServiceProvider implements ServiceProviderInterface
+{
+    /**
+     * @var string The URL for the Codebender website
+     */
+    protected $codebender_url = 'https://codebender.cc';
```
Comment (fpapadopou) [inline src/services/CodebenderServiceProvider.php]: One more thing to put in your config. The codebender url is not going to change, but it's a good practice to keep such thing outside your code.

File: app/Blocklyduino.php (hunk 1 above)
Comment (kyleoliveira) [inline app/Blocklyduino.php]: Ah, good idea. We'll do that.

File: app/Blocklyduino.php (hunk 2 above)
Comment (kyleoliveira) [inline app/Blocklyduino.php]: That's a good idea too. We can do that.

File: src/services/CodebenderServiceProvider.php (hunk 3 above)
Comment (kyleoliveira) [inline src/services/CodebenderServiceProvider.php]: Sure, why not?

File: app/Blocklyduino.php (hunk 1 above)
Comment (fpapadopou) [inline app/Blocklyduino.php]: :+1:

File: app/Blocklyduino.php (hunk 2 above)
Comment (fpapadopou) [inline app/Blocklyduino.php]: :+1:

File: src/services/CodebenderServiceProvider.php (hunk 3 above)
Comment (fpapadopou) [inline src/services/CodebenderServiceProvider.php]: :+1:

**Comment by fpapadopou** [comment]
@kyleoliveira could you also create a "default" settings.yml file? You could add some dummy settings, just for reference.

**Comment by kyleoliveira** [comment]
@fpapadopou Does the stuff in 0bd9141 and 2278644 finish this off?

File: src/views/blockly.html.twig [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+
+        /**
+         * Redefines compilerflasher's getbin function so that it points at our builder route.
+         * @param callback The callback function that hooks into getbin.
+         */
+        compilerflasher.getbin = function(callback) {
+            window.operationInProgress = true;
+            var payload = this.generate_payload("binary");
+            var cb = this;
+            $.post("\x2Fbuilder\x2Fcompile", payload, function (data) {
```
Comment (fpapadopou) [inline src/views/blockly.html.twig]: Could this be replaced with an asset that generates the compile route url?

**Comment by fpapadopou** [comment]
Can we remove the event reports from the blocklyduino page? There is no need to know which event was fired.

**Comment by fpapadopou** [comment]
@kyleoliveira @azyth Other than these two comments, there is nothing more for this pull request. However, checking out to this branch on blockly.codebender.cc made the website return a 500 error. Is this related to the changes in the code?

**Comment by kyleoliveira** [comment]
@fpapadopou I think I've addressed your last two comments with 93aafe9 . I also threw in 5912e7a because the quickstart.sh script should be able to point at the master branch once this gets merged in.

With respect to the blockly.codebender.cc site... again, it would probably be best for you to not poke that thing on your own... The issues you encountered are due to the addition of the YAML provider in the new commits (if you check out the error log at /var/log/nginx/error.log, you'll probably see some complaints about missing classes). When you did whatever you did, you did not rerun `sudo php ~/composer.phar update` and so, naturally, the supplied class wasn't available in the vendor folder and php got confused and you got a whitescreen. It seems like every time we deploy to the AWS site, particularly on this task, we've needed to do this composer update to make sure the packages are all in place. I've also gotten into the habit of restarting the php5-fpm service when I do this kind of thing, just to be safe.

If you want the blockly.codebender.cc site to be consistently available, I'd suggest two things:

1) Perhaps we can set up a dev.blockly.codebender.cc site where we are more tolerant of breaking things that can track a separate branch, while we leave the current site on master?

2) Can we explore using Docker for this site? We can set up a Docker repo that has the whole project ready to rock after every repo commit and then use the Docker repo for the deployment instead of the two step process of pulling from git and (potentially) rerunning composer. It would also open up some other hosting options if you guys are interested in that sort of thing. Just throwing that out there, but I'd understand if you haven't messed with this before and you're leery about it.

**Comment by fpapadopou** [comment]
@kyleoliveira @azyth I consider this pull request complete. I'm closing it and merging to master.

Regarding @kyleoliveira's comments:
Sorry for messing the AWS machine. I'm not sure we really need a dev AWS VM. The blockly.codebender.cc is meant to be there for development purposes, so we can be tolerant about connecting to it and running composer or restarting PHP or whatever. Which is definitely not right for a production environment. However, this brings me to your second comment, which I find a great idea. It will be one of your next tasks, after a discussion with the rest of the team.

**Pull request merged**